easyedge-accelerometer-example/easy-edge-firmware/components/astarte_handler/astarteHandler.c

```diff
@@ -0,0 +1,214 @@
+    .name = "org.astarteplatform.esp32.ServerDatastream",
+    .major_version = 0,
+    .minor_version = 1,
+    .ownership = OWNERSHIP_SERVER,
+    .type = TYPE_DATASTREAM,
+};
+
+static void astarte_data_events_handler(astarte_device_data_event_t *event)
+{
+    ESP_LOGI(TAG, "Got Astarte data event, interface_name: %s, path: %s, bson_type: %d",
+        event->interface_name, event->path, event->bson_value_type);
+
+    if (strcmp(event->interface_name, server_datastream_interface.name) == 0
+        && strcmp(event->path, "/led") == 0 && event->bson_value_type == BSON_TYPE_BOOLEAN) {
+        int led_state = astarte_bson_value_to_int8(event->bson_value);
+
+        esp_event_post(ASTARTE_EVENTS, ASTARTE_EVENT_LED, (void*) &led_state, sizeof(int), 0);
+    }
+}
+
+static void astarte_connection_events_handler()
+{
+    esp_event_post(ASTARTE_EVENTS, ASTARTE_EVENT_CONNECT, NULL, 0, 0);
+    ESP_LOGI(TAG, "on_connected");
+}
+
+static void astarte_disconnection_events_handler()
+{
+    esp_event_post(ASTARTE_EVENTS, ASTARTE_EVENT_CONNECT, NULL, 0, 0);
+    ESP_LOGW(TAG, "on_disconnected");
+}
+
+esp_err_t add_interfaces(astarte_device_handle_t device)
+{
+    astarte_err_t ret;
+    ret = astarte_device_add_interface(device, &acc_data_interface);
+    ret |= astarte_device_add_interface(device, &server_datastream_interface);
+
+    if (ret == ASTARTE_OK) {
+        ESP_LOGI(TAG, "ADDING INTERFACES OK");
+        return ESP_OK;
+    } else {
+        ESP_LOGE(TAG, "ERROR ADDING INTERFACES");
+        return ESP_FAIL;
+    }
+}
+
+AstarteHandler *astarteHandler_init()
+{
+    AstarteHandler *astarteHandler = NULL;
+    astarte_device_handle_t device = NULL;
+    char *encoded_hwid = NULL;
+
+    astarte_credentials_use_nvs_storage(NVS_PARTITION);
+    astarte_credentials_init();
+
+    encoded_hwid = astarteHandler_getHardwareIdEncoded();
+
+    if (encoded_hwid == NULL) {
+        goto astarte_init_error;
+    }
+
+    ESP_LOGI(TAG, "Astarte Device ID -> %s", encoded_hwid);
+    astarte_device_config_t cfg = { .data_event_callback = astarte_data_events_handler,
+        .connection_event_callback = astarte_connection_events_handler,
+        .disconnection_event_callback = astarte_disconnection_events_handler,
+        .hwid = (char *) encoded_hwid };
+
+    device = astarte_device_init(&cfg);
+    free(encoded_hwid);
+
+    if (!device) {
+        ESP_LOGE(TAG, "Cannot to init astarte device");
+        goto astarte_init_error;
+    }
+
+    ESP_ERROR_CHECK(add_interfaces(device));
+    astarteHandler = malloc(sizeof(AstarteHandler));
+
+    if (!astarteHandler) {
+        ESP_LOGE(TAG, "Cannot allocate astarte handler");
+        goto astarte_init_error;
+    }
+
+    astarteHandler->astarteDeviceHandle = device;
+    astarteHandler->start = &start;
+    astarteHandler->stop = &stop;
+
+    return astarteHandler;
+
+astarte_init_error:
+    if (device)
+        astarte_device_destroy(device);
```

Comment:
we should avoid `if` without braces, for a number of reasons the following is better:
```c
if (device) {
    astarte_device_destroy(device);
}
```